crates/extensions/tedge_mqtt_bridge/src/lib.rs

```diff
@@ -0,0 +1,202 @@
+                .await
+                .unwrap();
+        }
+
+        let (tx_pubs_from_cloud, rx_pubs_from_cloud) = mpsc::channel(10);
+        let (tx_pubs_from_local, rx_pubs_from_local) = mpsc::channel(10);
+        tokio::spawn(one_way_bridge(
+            local_event_loop,
+            cloud_client,
+            move |topic| topic.strip_prefix(&topic_prefix).unwrap().into(),
+            tx_pubs_from_local,
+            rx_pubs_from_cloud,
+        ));
+        tokio::spawn(one_way_bridge(
+            cloud_event_loop,
+            local_client,
+            move |topic| format!("{prefix}/{topic}").into(),
+            tx_pubs_from_cloud,
+            rx_pubs_from_local,
+        ));
+
+        Self { signal_sender }
+    }
+
+    pub(crate) fn build_actor(self) -> MqttBridgeActor {
+        MqttBridgeActor {}
+    }
+}
+
+async fn one_way_bridge<F: for<'a> Fn(&'a str) -> Cow<'a, str>>(
+    mut recv_event_loop: EventLoop,
+    target: AsyncClient,
+    transform_topic: F,
+    mut tx_pubs: mpsc::Sender<Publish>,
+    mut rx_pubs: mpsc::Receiver<Publish>,
+) {
+    let mut forward_pkid_to_received_msg = HashMap::new();
+    let mut last_err = None;
+    loop {
+        let notification = match recv_event_loop.poll().await {
+            // TODO notify if this is us recovering from an error
+            Ok(notification) => notification,
+            Err(err) => {
+                let err = err.to_string();
+                if last_err.as_ref() != Some(&err) {
+                    // TODO clarify whether this is cloud/local
+                    error!("MQTT bridge connection error: {err}");
+                    last_err = Some(err);
+                }
+                continue;
+            }
+        };
+        match notification {
+            // Forwarding messages from event loop to target
+            Event::Incoming(Incoming::Publish(publish)) => {
+                target
+                    .publish(
+                        transform_topic(&publish.topic),
+                        publish.qos,
+                        publish.retain,
+                        publish.payload.clone(),
+                    )
+                    .await
+                    .unwrap();
+                tx_pubs.send(publish).await.unwrap();
+            }
+            // Forwarding acks from event loop to target
+            Event::Incoming(
+                Incoming::PubAck(PubAck { pkid: ack_pkid })
+                | Incoming::PubRec(PubRec { pkid: ack_pkid }),
+            ) => {
+                target
+                    .ack(&forward_pkid_to_received_msg.remove(&ack_pkid).unwrap())
+                    .await
+                    .unwrap();
+            }
+            Event::Outgoing(Outgoing::Publish(pkid)) => {
+                if let Some(msg) = rx_pubs.next().await {
+                    forward_pkid_to_received_msg.insert(pkid, msg);
+                } else {
+                    break;
+                }
+            }
```

Comment:
Is this really robust, as this requires a perfect synchronization between two event streams? Sure these streams are not independent (the companion loop triggers one event for this `rx_pubs` and publishes a message triggering  the `OutGoing` event. But I would be cautious here. Mainly because of QoS subtleties. Do we have such an outgoing event for QoS 0? If not, then no message should be propagated to the companion loop.

Reply:
I replied to this mostly in https://github.com/thin-edge/thin-edge.io/pull/2716#discussion_r1517673228. I think the answer is yes, we are robust as QoS 0 messages do go through the same series of events, just the packet identifiers are all 0.

Reply:
> just the packet identifiers are all 0

Okay then